File: src/test/java/tech/sirwellington/alchemy/http/mock/MockStepsTest.java
/*
 * Copyright © 2018. Sir Wellington.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.http.mock;

import java.net.URL;
import java.time.Instant;
import java.util.*;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import tech.sirwellington.alchemy.http.AlchemyRequestSteps;
import tech.sirwellington.alchemy.http.AlchemyRequestSteps.OnFailure;
import tech.sirwellington.alchemy.http.AlchemyRequestSteps.OnSuccess;
import tech.sirwellington.alchemy.http.HttpResponse;
import tech.sirwellington.alchemy.http.exceptions.AlchemyHttpException;
import tech.sirwellington.alchemy.http.mock.MockSteps.*;
import tech.sirwellington.alchemy.test.junit.runners.*;

import static org.hamcrest.Matchers.*;
import static org.junit.Assert.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.*;
import static tech.sirwellington.alchemy.generator.AlchemyGenerator.Get.one;
import static tech.sirwellington.alchemy.generator.NetworkGenerators.httpUrls;
import static tech.sirwellington.alchemy.generator.StringGenerators.strings;
import static tech.sirwellington.alchemy.test.junit.ThrowableAssertion.*;

/**
 *
 * @author SirWellington
 */
@Repeat(10)
@RunWith(AlchemyTestRunner.class)
public class MockStepsTest
{

    @Mock
    private MockAlchemyHttp mockHttp;

    @Mock
    private OnSuccess<HttpResponse> successCallback;

    @Mock
    private OnFailure failureCallback;

    @GeneratePojo
    private MockRequest request;

    @GenerateDate
    private Date date;

    @Mock
    private HttpResponse httpResponse;

    @Before
    public void setUp()
    {

        when(mockHttp.getResponseFor(request))
            .thenReturn(httpResponse);
    }

    @DontRepeat
    @Test
    public void testCannotInstantiateClass()
    {
        assertThrows(() -> new MockSteps());
    }

    @DontRepeat
    @Test
    public void testStep1Constructor()
    {
        assertThrows(() -> new MockStep1(null)).isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void testStep1Get()
    {
        MockStep1 step1 = new MockStep1(mockHttp);

        AlchemyRequestSteps.Step3 get = step1.get();
        assertThat(get, notNullValue());
        assertThat(get, is(instanceOf(MockStep3.class)));
        MockStep3 mockStep3 = (MockStep3) get;
        assertThat(mockStep3.getRequest().method, is(MockRequest.Method.GET));
    }

    @Test
    public void testStep1Post()
    {
        MockStep1 step1 = new MockStep1(mockHttp);

        AlchemyRequestSteps.Step2 post = step1.post();
        assertThat(post, notNullValue());
        assertThat(post, is(instanceOf(MockStep2.class)));
        MockStep2 mockStep2 = (MockStep2) post;
        assertThat(mockStep2.getRequest().method, is(MockRequest.Method.POST));
    }

    @Test
    public void testStep1Put()
    {
        MockStep1 step1 = new MockStep1(mockHttp);

        AlchemyRequestSteps.Step2 put = step1.put();
        assertThat(put, notNullValue());
        assertThat(put, is(instanceOf(MockStep2.class)));
        MockStep2 mockStep2 = (MockStep2) put;
        assertThat(mockStep2.getRequest().method, is(MockRequest.Method.PUT));
    }

    @Test
    public void testStep1Delete()
    {
        MockStep1 step1 = new MockStep1(mockHttp);

        AlchemyRequestSteps.Step2 delete = step1.delete();
        assertThat(delete, notNullValue());
        assertThat(delete, is(instanceOf(MockStep2.class)));
        MockStep2 mockStep2 = (MockStep2) delete;
        assertThat(mockStep2.getRequest().method, is(MockRequest.Method.DELETE));
    }

    @DontRepeat
    @Test
    public void testStep2Constructor()
    {
        assertThrows(() -> new MockStep2(null, request))
            .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> new MockStep2(mockHttp, null))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void testStep2WithNoBody()
    {
        MockStep2 step2 = new MockStep2(mockHttp, request);

        AlchemyRequestSteps.Step3 nothing = step2.nothing();
        assertThat(nothing, notNullValue());
        assertThat(nothing, is(instanceOf(MockStep3.class)));

        MockStep3 step3 = (MockStep3) nothing;
        assertThat(step3.getMockAlchemyHttp(), is(mockHttp));
        assertThat(step3.getRequest().method, is(request.method));
        assertThat(step3.getRequest().body, is(MockRequest.NO_BODY));
    }

    @Test
    public void testStep2WithPojo()
    {
        MockStep2 step2 = new MockStep2(mockHttp, request);

        AlchemyRequestSteps.Step3 body = step2.body(date);
        assertThat(body, notNullValue());
        assertThat(body, is(instanceOf(MockStep3.class)));

        MockStep3 step3 = (MockStep3) body;
        assertThat(step3.getMockAlchemyHttp(), is(mockHttp));
        assertThat(step3.getRequest().method, is(request.method));
        assertThat(step3.getRequest().body, is(date));
    }

    @DontRepeat
    @Test
    public void testStep2WithNullPojo()
    {
        MockStep2 step2 = new MockStep2(mockHttp, request);

        assertThrows(() -> step2.body(null))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void testStep2WithString()
    {
        MockStep2 step2 = new MockStep2(mockHttp, request);

        String string = one(strings());
        AlchemyRequestSteps.Step3 body = step2.body(string);
        assertThat(body, notNullValue());
        assertThat(body, is(instanceOf(MockStep3.class)));

        MockStep3 step3 = (MockStep3) body;
        assertThat(step3.getMockAlchemyHttp(), is(mockHttp));
        assertThat(step3.getRequest().method, is(request.method));
        assertThat(step3.getRequest().body, is(string));
    }

    @DontRepeat
    @Test
    public void testStep3Constructor()
    {
        assertThrows(() -> new MockStep3(mockHttp, null))
            .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> new MockStep3(null, request))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void testStep3()
    {
        MockStep3 step3 = new MockStep3(mockHttp, request);

    }

    @Test
    public void testStep3Expecting()
    {
        MockStep3 step3 = new MockStep3(mockHttp, request);
        List<Class<?>> classes = Arrays.asList(String.class, Date.class, Instant.class);
        Class<?> expected = classes.stream().findAny().get();

        AlchemyRequestSteps.Step4<?> expecting = step3.expecting(expected);
        assertThat(expecting, notNullValue());
        assertThat(expecting, is(instanceOf(MockStep4.class)));

        MockStep4<?> step4 = (MockStep4<?>) expecting;
        assertThat(step4.getExpectedClass(), sameInstance(expected));
        assertThat(step4.getMockAlchemyHttp(), is(mockHttp));
        assertThat(step4.getRequest(), is(request));
    }

    @Test
    public void testStep3At()
    {
        MockStep3 step3 = new MockStep3(mockHttp, request);
        URL url = one(httpUrls());

        when(mockHttp.getResponseFor(request))
            .thenReturn(httpResponse);

        HttpResponse response = step3.at(url);
        assertThat(request.url, is(url));
        verify(mockHttp).getResponseFor(request);

        assertThrows(() -> step3.at((URL) null))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void testStep4()
    {
        MockStep4<HttpResponse> step4 = new MockStep4<>(mockHttp, request, HttpResponse.class);

        AlchemyRequestSteps.Step5 response = step4.onSuccess(successCallback);

        assertThat(response, notNullValue());
        assertThat(response, is(instanceOf(MockSteps.MockStep5.class)));

        MockSteps.MockStep5 step5 = (MockSteps.MockStep5) response;
        assertThat(step5.getMockAlchemyHttp(), is(mockHttp));
        assertThat(step5.getExpectedClass(), equalTo(HttpResponse.class));
        assertThat(step5.getOnSuccessCallback(), is(successCallback));
        assertThat(step5.getRequest(), is(request));

    }

    @DontRepeat
    @Test
    public void testStep4WithBadArgs()
    {
        assertThrows(() -> new MockStep4<>(null, request, HttpResponse.class));
        assertThrows(() -> new MockStep4<>(mockHttp, null, HttpResponse.class));
        assertThrows(() -> new MockStep4<>(mockHttp, request, null));

        MockStep4<HttpResponse> instance = new MockStep4<>(mockHttp, request, HttpResponse.class);
        assertThrows(() -> instance.onSuccess(null));
    }

    @Test
    public void testStep5()
    {

        MockSteps.MockStep5<HttpResponse> instance = new MockSteps.MockStep5<>(mockHttp, successCallback, HttpResponse.class, request);
        AlchemyRequestSteps.Step6<HttpResponse> response = instance.onFailure(failureCallback);
        assertThat(response, notNullValue());
        assertThat(response, is(instanceOf(MockSteps.MockStep6.class)));

        MockSteps.MockStep6 step6 = (MockSteps.MockStep6) response;
        assertThat(step6.getExpectedClass(), equalTo(HttpResponse.class));
        assertThat(step6.getMockAlchemyHttp(), is(mockHttp));
        assertThat(step6.getOnSuccessCallback(), is(successCallback));
        assertThat(step6.getOnFailureCallback(), is(failureCallback));
        assertThat(step6.getRequest(), is(request));
    }

    @DontRepeat
    @Test
    public void testStep5WithBadArgs()
    {
        assertThrows(() -> new MockSteps.MockStep5<>(null, OnSuccess.Companion.getNO_OP(), Object.class, request));
        assertThrows(() -> new MockSteps.MockStep5<>(mockHttp, null, String.class, request));
        assertThrows(() -> new MockSteps.MockStep5<>(mockHttp, AlchemyRequestSteps.OnSuccess.INSTANCES.NO_OP, null, request));
        assertThrows(() -> new MockSteps.MockStep5<>(mockHttp, AlchemyRequestSteps.OnSuccess.INSTANCES.NO_OP, Object.class, null));

        MockSteps.MockStep5<Object> instance = new MockSteps.MockStep5<>(mockHttp, OnSuccess.INSTANCES.NO_OP, Object.class, request);
        assertThrows(() -> instance.onFailure(null))
            .isInstanceOf(IllegalArgumentException.class);

    }

    @Test
    public void testStep6()
    {
        MockSteps.MockStep6 instance = new MockSteps.MockStep6(mockHttp,
                                                               successCallback,
                                                               failureCallback,
                                                               HttpResponse.class,
                                                               request);

        when(mockHttp.getResponseFor(request, HttpResponse.class))
            .thenReturn(httpResponse);

        instance.at(request.url);
        verify(mockHttp).getResponseFor(request, HttpResponse.class);
        verify(successCallback).processResponse(httpResponse);
    }

    @Test
    public void testStep6WhenFails()
    {

        when(mockHttp.getResponseFor(request, HttpResponse.class))
            .thenThrow(new AlchemyHttpException());

        MockSteps.MockStep6 instance = new MockSteps.MockStep6(mockHttp,
                                                               successCallback,
                                                               failureCallback,
                                                               HttpResponse.class,
                                                               request);

        instance.at(request.url);
        verifyZeroInteractions(successCallback);
        verify(failureCallback).handleError(any());
    }

    @DontRepeat
    @Test
    public void testStep6WithBadArgs()
    {
        assertThrows(() -> new MockSteps.MockStep6(null, successCallback, failureCallback, HttpResponse.class, request));
        assertThrows(() -> new MockSteps.MockStep6(mockHttp, null, failureCallback, HttpResponse.class, request));
        assertThrows(() -> new MockSteps.MockStep6(mockHttp, successCallback, null, HttpResponse.class, request));
        assertThrows(() -> new MockSteps.MockStep6(mockHttp, successCallback, failureCallback, null, request));
        assertThrows(() -> new MockSteps.MockStep6(mockHttp, successCallback, failureCallback, HttpResponse.class, null));

        MockSteps.MockStep6 instance = new MockSteps.MockStep6(mockHttp, successCallback, failureCallback, HttpResponse.class, request);
        assertThrows(() -> instance.at((URL) null))
            .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> instance.at((String) null))
            .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> instance.at(""))
            .isInstanceOf(IllegalArgumentException.class);


    }
}

File: src/test/java/tech/sirwellington/alchemy/http/mock/MockStepsTest.kt
/*
 * Copyright © 2019. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.http.mock

import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import com.natpryce.hamkrest.isA
import com.natpryce.hamkrest.present
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Answers
import org.mockito.Mock
import org.mockito.Mockito.verifyZeroInteractions
import sir.wellington.alchemy.collections.lists.Lists
import tech.sirwellington.alchemy.generator.AlchemyGenerator.Get.one
import tech.sirwellington.alchemy.generator.NetworkGenerators.Companion.httpUrls
import tech.sirwellington.alchemy.generator.StringGenerators.Companion.strings
import tech.sirwellington.alchemy.http.AlchemyRequestSteps.OnFailure
import tech.sirwellington.alchemy.http.AlchemyRequestSteps.OnSuccess
import tech.sirwellington.alchemy.http.HttpResponse
import tech.sirwellington.alchemy.http.RequestMethod
import tech.sirwellington.alchemy.http.exceptions.AlchemyHttpException
import tech.sirwellington.alchemy.http.mock.MockSteps.MockStep1
import tech.sirwellington.alchemy.http.mock.MockSteps.MockStep2
import tech.sirwellington.alchemy.http.mock.MockSteps.MockStep3
import tech.sirwellington.alchemy.http.mock.MockSteps.MockStep4
import tech.sirwellington.alchemy.http.mock.MockSteps.MockStep5
import tech.sirwellington.alchemy.http.mock.MockSteps.MockStep6
import tech.sirwellington.alchemy.test.junit.ThrowableAssertion.assertThrows
import tech.sirwellington.alchemy.test.junit.runners.*
import java.net.URL
import java.time.Instant
import java.util.Date

/**
 *
 * @author SirWellington
 */
@Repeat(10)
@RunWith(AlchemyTestRunner::class)
class MockStepsTest
{

    @Mock(answer = Answers.RETURNS_MOCKS)
    private lateinit var  mockHttp: MockAlchemyHttp

    @Mock
    private lateinit var  successCallback: OnSuccess<HttpResponse>

    @Mock
    private lateinit var  failureCallback: OnFailure

    @GeneratePojo
    private lateinit var  request: MockRequest

    @GenerateDate
    private lateinit var  date: Date

    @Mock
    private lateinit var httpResponse: HttpResponse

    @Before
    fun setUp()
    {
        whenever(mockHttp.getResponseFor(request))
                .thenReturn(httpResponse)
    }

    @DontRepeat
    @Test
    fun testCannotInstantiateClass()
    {
        assertThrows { MockSteps::class.java.newInstance() }
    }

    @Test
    fun testStep1Get()
    {
        val step1 = MockStep1(mockHttp)

        val get = step1.get()
        assertThat(get, present())
        assertThat(get, isA<MockStep3>())
        val mockStep3 = get as MockStep3
        assertThat(mockStep3.request.method, present(equalTo(RequestMethod.GET)))
    }

    @Test
    fun testStep1Post()
    {
        val step1 = MockStep1(mockHttp)

        val post = step1.post()
        assertThat(post, present())
        assertThat(post, isA<MockStep2>())
        val mockStep2 = post as MockStep2
        assertThat(mockStep2.request.method, equalTo(RequestMethod.POST))
    }

    @Test
    fun testStep1Put()
    {
        val step1 = MockStep1(mockHttp)

        val put = step1.put()
        assertThat(put, present())
        assertThat(put, isA<MockStep2>())
        val mockStep2 = put as MockStep2
        assertThat(mockStep2.request.method!!, equalTo(RequestMethod.PUT))
    }

    @Test
    fun testStep1Delete()
    {
        val step1 = MockStep1(mockHttp)

        val delete = step1.delete()
        assertThat(delete, present())
        assertThat(delete, isA<MockStep2>())
        val mockStep2 = delete as MockStep2
        assertThat(mockStep2.request.method, equalTo(RequestMethod.DELETE))
    }

    @Test
    fun testStep2WithNoBody()
    {
        val step2 = MockStep2(mockHttp, request)

        val nothing = step2.nothing()
        assertThat(nothing, present())
        assertThat(nothing, isA<MockStep3>())

        val step3 = nothing as MockStep3
        assertThat(step3.mockAlchemyHttp, equalTo(mockHttp))
        assertThat(step3.request.method, equalTo(request.method))
        assertThat(step3.request.body, equalTo(MockRequest.NO_BODY))
    }

    @Test
    fun testStep2WithPojo()
    {
        val step2 = MockStep2(mockHttp, request)

        val body = step2.body(date)
        assertThat(body, present())
        assertThat(body, isA<MockStep3>())

        val step3 = body as MockStep3
        assertThat(step3.mockAlchemyHttp, equalTo(mockHttp))
        assertThat(step3.request.method, equalTo(request.method))
        assertThat(step3.request.body, present<Any>(equalTo(date)))
    }


    @Test
    fun testStep2WithString()
    {
        val step2 = MockStep2(mockHttp, request)

        val string = one<String>(strings())
        val body = step2.body(string)
        assertThat(body, present())
        assertThat(body, isA<MockStep3>())

        val step3 = body as MockStep3
        assertThat(step3.mockAlchemyHttp, equalTo(mockHttp))
        assertThat(step3.request.method, equalTo(request.method))
        assertThat(step3.request.body, present<Any>(equalTo(string)))
    }

    @Test
    fun testStep3()
    {
        val step3 = MockStep3(mockHttp, request)

    }

    @Test
    fun testStep3Expecting()
    {
        val step3 = MockStep3(mockHttp, request)
        val classes = Lists.createFrom(String::class.java, Date::class.java, Instant::class.java)
        val expected = classes.stream().findAny().get()

        val expecting = step3.expecting(expected)
        assertThat(expecting, present())
        assertThat(expecting, isA<MockStep4<*>>())

        val step4 = expecting as MockStep4<*>
        assertThat(step4.expectedClass, present<Class<*>>(equalTo(expected)))
        assertThat(step4.mockAlchemyHttp, equalTo(mockHttp))
        assertThat(step4.request, equalTo(request))
    }

    @Test
    fun testStep3At()
    {
        val step3 = MockStep3(mockHttp, request)
        val url = one<URL>(httpUrls())

        whenever(mockHttp.getResponseFor(request))
                .thenReturn(httpResponse)

        val response = step3.at(url)
        assertThat(response, equalTo(httpResponse))
        assertThat(request.url, equalTo(url))
        verify(mockHttp).getResponseFor(request)

    }

    @Test
    fun testStep4()
    {
        val step4 = MockStep4(mockHttp, request, HttpResponse::class.java)

        val response = step4.onSuccess(successCallback)

        assertThat(response, present())
        assertThat(response, isA<MockStep5<*>>())

        val step5 = response as MockSteps.MockStep5<*>
        assertThat(step5.mockAlchemyHttp, equalTo(mockHttp))
        assertThat(step5.expectedClass, present<Class<*>>(equalTo(HttpResponse::class.java)))
        assertThat(step5.onSuccessCallback, present<Any>(equalTo(successCallback)))
        assertThat(step5.request, equalTo(request))

    }

    @Test
    fun testStep5()
    {

        val instance = MockSteps.MockStep5(mockHttp, successCallback, HttpResponse::class.java, request)
        val response = instance.onFailure(failureCallback)
        assertThat(response, present())
        assertThat(response, isA<MockStep6<*>>())

        val step6 = response as MockSteps.MockStep6<*>
        assertThat(step6.expectedClass, equalTo<Class<*>>(HttpResponse::class.java))
        assertThat(step6.mockAlchemyHttp, equalTo(mockHttp))
        assertThat(step6.onSuccessCallback, present<Any>(equalTo(successCallback)))
        assertThat(step6.onFailureCallback, equalTo(failureCallback))
        assertThat(step6.request, equalTo(request))
    }

    @Test
    fun testStep6()
    {
        val instance = MockSteps.MockStep6(mockHttp,
                                           successCallback,
                                           failureCallback,
                                           HttpResponse::class.java,
                                           request)

        whenever(mockHttp.getResponseFor(request, HttpResponse::class.java))
                .thenReturn(httpResponse)

        instance.at(request.url!!)
        verify(mockHttp).getResponseFor(request, HttpResponse::class.java)
        verify(successCallback).processResponse(httpResponse)
    }

    @Test
    fun testStep6WhenFails()
    {

        whenever(mockHttp.getResponseFor(request, HttpResponse::class.java))
                .thenThrow(AlchemyHttpException())

        val instance = MockSteps.MockStep6(mockHttp,
                                           successCallback,
                                           failureCallback,
                                           HttpResponse::class.java,
                                           request)

        instance.at(request.url!!)
        verifyZeroInteractions(successCallback)
        verify(failureCallback).handleError(any())
    }

    @DontRepeat
    @Test
    fun testStep6WithBadArgs()
    {
        val instance = MockSteps.MockStep6(mockHttp, successCallback, failureCallback, HttpResponse::class.java, request)

        assertThrows { instance.at("") }
                .isInstanceOf(IllegalArgumentException::class.java)


    }
}
